Add idempotency check to prevent duplicate `InvoiceReceived` events

shaavan opened the pull request:
Solves #3653

When `manually_handle_bolt12_invoice` is enabled, we shouldn’t emit multiple `InvoiceReceived` events for the same invoice — it can be confusing and misleading for consumers of the event stream.

This PR adds a simple idempotency check to make sure we only emit the event once per invoice.

Also added a test to make sure this behavior stays in place going forward.

shaavan commented on lightning/src/ln/channelmanager.rs:
@benthecarman

> it seems send_payment_for_bolt12_invoice is not idempotent either so this would effect everyone using manually_handle_bolt12_invoices

Hi Ben! AFAIU, `send_payment_for_bolt12_invoice` does seem idempotent, as in, it won’t allow paying the same invoice twice (see [this link](https://github.com/lightningdevkit/rust-lightning/pull/3658/files#diff-36a745bde2c06cd0ac31c78125baeb280fe3ec60a660b9c1dfa225d9c0a49264R1210-R1214)).

Let me know if I’m misunderstanding the issue or if there’s still a potential problem here that I might be missing. Thanks!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -12477,6 +12477,23 @@ where
 
 				if self.default_configuration.manually_handle_bolt12_invoices {
+					// Prevent duplicate `InvoiceReceived` events by checking if this invoice was already processed.
+					let already_present = self.pending_events.lock().unwrap().iter().any(|(e, _)| {
+						matches!(e, Event::InvoiceReceived { payment_id: existing_id, .. } if existing_id == &payment_id)
+					});
```
Can't we do a lookup in `PendingOutboundPayments` by payment id and check it's state? We could transition it to the `InvoiceReceived` state before generating the event.

shaavan commented:
**Updated from** [pr3658.01](https://github.com/shaavan/rust-lightning/commits/pr3658.01) to [pr3658.02](https://github.com/shaavan/rust-lightning/commits/pr3658.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3658.01..pr3658.02)):
Addressed @jkczyz (offline) comments

Changes:
1. Update the approach, such that we update the `PendingOutboundPayment` entry the moment we receive the invoice to ensure event generation idempotency.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -12477,6 +12478,25 @@ where
 				);
 
 				if self.default_configuration.manually_handle_bolt12_invoices {
+					// Update the corresponding entry in `PendingOutboundPayment` for this invoice.
+					// This ensures that event generation remains idempotent in case we receive
+					// the same invoice multiple times.
+					match self.pending_outbound_payments.pending_outbound_payments.lock().unwrap().entry(payment_id) {
+						hash_map::Entry::Occupied(entry) => match entry.get() {
+							PendingOutboundPayment::AwaitingInvoice {
+								retry_strategy, route_params_config, ..
+							} => {
+								*entry.into_mut() = PendingOutboundPayment::InvoiceReceived {
+									payment_hash: invoice.payment_hash(),
+									retry_strategy: *retry_strategy,
+									route_params_config: *route_params_config,
+								};
+							},
+							_ => return None
+						},
+						hash_map::Entry::Vacant(_) => return None,
+					}
```
Let's make a method on `PendingOutboundPayment` instead so the behavior is encapsulated there.

shaavan replied:
**Done!** [pr3658.03](https://github.com/shaavan/rust-lightning/commits/pr3658.03) 
I’ve restructured the functions for better separation of logic and to allow us to encapsulate the added code within `ChannelManager`.  
Thanks a lot for the suggestion! ✨

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1789,6 +1772,54 @@ impl OutboundPayments {
+	fn mark_invoice_received(
+		&self, payment_id: PaymentId, payment_hash: PaymentHash, with_manual_handling: bool
+	) -> Result<(Retry, RouteParametersConfig), Bolt12PaymentError> {
+		let retry_strategy;
+		let params_config;
```
It would be cleaner to simply return directly instead of needing to assign to these first.

shaavan replied:
Done! [pr3658.03](https://github.com/shaavan/rust-lightning/commits/pr3658.03)
Thanks for the pointer!

jkczyz commented:
@TheBlueMatt Is the following line still correct after this change?

https://github.com/lightningdevkit/rust-lightning/blob/cfdd217d16fedaee512b6d6bc45b27a1db37e59a/lightning/src/ln/outbound_payment.rs#L2362

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1789,6 +1772,54 @@ impl OutboundPayments {
+			// current one is handled by the user.
+			match self.pending_outbound_payments.lock().unwrap().entry(payment_id) {
+				hash_map::Entry::Occupied(entry) => match entry.get() {
+					PendingOutboundPayment::InvoiceReceived {
```
Wouldn't the second lock acquisition happening in `send_payment_for_bolt12_invoice_internal` prevent that? We'd still attempt to find a path twice but would fail because the first attempt transitions the state to `Retryable` causing the second attempt to return `DuplicateInvoice`.

shaavan commented:
**Updated from** [pr3658.02](https://github.com/shaavan/rust-lightning/commits/pr3658.02) to [pr3658.03](https://github.com/shaavan/rust-lightning/commits/pr3658.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3658.02..pr3658.03)):
Addressed @jkczyz comments

Changes:
1. Restructure functions, to have better logical seperation.
2. Encapsulate `handle_message` code to improve code readability.

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -863,25 +863,14 @@ impl OutboundPayments {
+		// is already updated at the time the invoice is received. This ensures that `InvoiceReceived`
+		// event generation remains idempotent, even if the same invoice is received again before the
+		// current one is handled by the user.
+		if !with_manual_handling { self.mark_invoice_received(payment_id, payment_hash)? }
+
+		let (retry_strategy, params_config) = self.received_invoice_details(payment_id)?;
```
Can we avoid taking the lock twice here when `with_manual_handling` is false?

shaavan replied:
Thanks for the pointer, Jeff!  
I’ve updated the `mark_invoice_received` function to also return the `retry` and `route_params`, so we can avoid taking the lock twice in [`pr3658.04`](https://github.com/shaavan/rust-lightning/commits/pr3658.04).  
Let me know if this looks good — thanks again!

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1789,6 +1778,44 @@ impl OutboundPayments {
+		}
+	}
+
+	fn received_invoice_details(
+		&self, payment_id: PaymentId,
+	) -> Result<(Retry, RouteParametersConfig), Bolt12PaymentError> {
+		match self.pending_outbound_payments.lock().unwrap().entry(payment_id) {
+			hash_map::Entry::Occupied(entry) => match entry.get() {
+				PendingOutboundPayment::InvoiceReceived {
+					retry_strategy, route_params_config, ..
+				} => {
+					return Ok((*retry_strategy, *route_params_config))
+				},
+				_ => return Err(Bolt12PaymentError::DuplicateInvoice),
+			},
+			hash_map::Entry::Vacant(_) => return Err(Bolt12PaymentError::UnexpectedInvoice),
+		}
+	}
```
Probably can re-inline this to avoid taking the lock again.

shaavan commented:
**Updated from** [pr3658.03](https://github.com/shaavan/rust-lightning/commits/pr3658.03) to [pr3658.04](https://github.com/shaavan/rust-lightning/commits/pr3658.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3658.03..pr3658.04)):
Addressed @jkczyz comments

Changes:
1. Update function signatures, and code to avoid taking lock twice.
2. Minor nit fixes

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -863,25 +863,16 @@ impl OutboundPayments {
+		// When manual invoice handling is enabled, the corresponding `PendingOutboundPayment` entry
+		// is already updated at the time the invoice is received. This ensures that `InvoiceReceived`
+		// event generation remains idempotent, even if the same invoice is received again before the
+		// current one is handled by the user.
```
s/current one/event

jkczyz reviewed (commented):
Looks good. Just some small changes.

shaavan commented:
**Updated from** [pr3658.04](https://github.com/shaavan/rust-lightning/commits/pr3658.04) to [pr3658.05](https://github.com/shaavan/rust-lightning/commits/pr3658.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3658.04..pr3658.05)):
Addressed @jkczyz comments

Changes:
1. Introduce new wrapper function mark_invoice_received for ChannelManager usage.
2. Update function signature to take invoice as input, and output `invoice.payment_hash()`
3. Update docs

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1789,6 +1780,46 @@ impl OutboundPayments {
 		}
 	}
 
+	pub(super) fn mark_invoice_received_and_get_details(
```
Addressed in [pr3658.05](https://github.com/shaavan/rust-lightning/commits/pr3658.05)
Thanks for the pointer! ✨

jkczyz reviewed (commented):
LGTM

shaavan commented:
**Updated from** [pr3658.05](https://github.com/shaavan/rust-lightning/commits/pr3658.05) to [pr3658.06](https://github.com/shaavan/rust-lightning/commits/pr3658.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3658.05..pr3658.06)):

Changes:
1. Squash commits.

shaavan commented:
**Updated from** [pr3658.06](https://github.com/shaavan/rust-lightning/commits/pr3658.06) to [pr3658.07](https://github.com/shaavan/rust-lightning/commits/pr3658.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3658.06..pr3658.07)):
Addressed @TheBlueMatt comments

Changes:
1. Remove the `manual_handling` parameter, and updated logic to _upgrade-if-necessay-and-get-details_.
2. Updated documentation to reflect that InvoiceReceived may now be written to the disk.

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -862,26 +862,15 @@ impl OutboundPayments {
+		// is already updated at the time the invoice is received. This ensures that `InvoiceReceived`
+		// event generation remains idempotent, even if the same invoice is received again before the
+		// event is handled by the user.
+		let (payment_hash, retry_strategy, params_config) = if with_manual_handling {
```
Makes sense! Thanks a lot for the pointer!
I have removed the manual_handling parameter in [pr3658.07](https://github.com/shaavan/rust-lightning/commits/pr3658.07), and update the logic, so that we update-if-necessary-and-get-details, when `get_received_invoice_details` is called.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -12477,6 +12478,11 @@ where
+					// Update the corresponding entry in `PendingOutboundPayment` for this invoice.
+					// This ensures that event generation remains idempotent in case we receive
+					// the same invoice multiple times.
+					self.pending_outbound_payments.mark_invoice_received(&invoice, payment_id).ok()?;
```
Thanks for catching it! I have updated the comment to reflect the changes in [pr3658.07](https://github.com/shaavan/rust-lightning/commits/pr3658.07)
Thanks!

shaavan commented:
**Updated from** [pr3658.07](https://github.com/shaavan/rust-lightning/commits/pr3658.07) to [pr3658.08](https://github.com/shaavan/rust-lightning/commits/pr3658.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3658.07..pr3658.08)):
Addressed @TheBlueMatt comments

Changes:
1. Squash commits
2. Renamed function and variable for improved clarity and better alignment with their actual purpose.

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1789,6 +1772,51 @@ impl OutboundPayments {
+			})
+	}
+
+	fn get_received_invoice_details(
```
I have updated the variable & function names in [pr3658.08](https://github.com/shaavan/rust-lightning/commits/pr3658.08)
Let me know if they look good! Thanks a lot for the pointer!
